python/whylogs/experimental/core/metrics/callback_metric.py

```diff
@@ -0,0 +1,121 @@
+    def resolve(self, name: str, why_type: DataType, fi_disabled: bool = False) -> Dict[str, Metric]:
+        metrics: Dict[str, Metric] = {
+            "counts": StandardMetric.counts.zero(MetricConfig()),
+            "types": StandardMetric.types.zero(MetricConfig()),
+            "cardinality": StandardMetric.cardinality.zero(MetricConfig()),
+        }
+
+        if isinstance(why_type, Integral):
+            metrics["distribution"] = StandardMetric.distribution.zero(MetricConfig())
+            metrics["ints"] = StandardMetric.ints.zero(MetricConfig())
+            if not fi_disabled:
+                metrics["frequent_items"] = StandardMetric.frequent_items.zero(MetricConfig())
+        elif isinstance(why_type, Fractional):
+            metrics["distribution"] = StandardMetric.distribution.zero(MetricConfig())
+        elif isinstance(why_type, String) and not fi_disabled:
+            metrics["frequent_items"] = StandardMetric.frequent_items.zero(MetricConfig())
+
+        return metrics
+
+
+@dataclass(frozen=True)
+class CallbackMetricConfig(MetricConfig):
+    """Maps feature name to callable that computes it"""
+
+    callbacks: Dict[str, Callable[[Any], Any]] = field(default_factory=dict)
+    submetric_schema: SubmetricSchema = field(default_factory=DefaultSchema)
+    type_mapper: TypeMapper = field(default_factory=StandardTypeMapper)
+
+
+class CallbackMetric(MultiMetric):
+    def __init__(
+        self,
+        callbacks: Dict[str, Callable[[Any], Any]],
+        # discover these with resolver  submetrics: Dict[str, Dict[str, Metric]],  # feature name -> (namespace -> metric)
+        submetric_schema: Optional[SubmetricSchema] = None,
+        type_mapper: Optional[TypeMapper] = None,
+        fi_disabled: bool = False,
+    ):
+        super().__init__(dict())  # submetrics)
+        self._callbacks = callbacks
+        self._submetric_schema = submetric_schema or DefaultSchema()
+        self._type_mapper = type_mapper or StandardTypeMapper()
+        self._fi_disabled = fi_disabled
+
+    @property
+    def namespace(self) -> str:
+        return "callback"
+
+    def merge(self, other: "CallbackMetric") -> "CallbackMetric":
+        merged = CallbackMetric(self._callbacks, self._submetric_schema, self._type_mapper, self._fi_disabled)
+        merged.submetrics = self.merge_submetrics(other)
+        return merged
+
+    def _add_submetric(self, name: str, value: Any) -> None:
+        self.submetrics[name] = self._submetric_schema.resolve(name, self._type_mapper(type(value)), self._fi_disabled)
+
+    def _update_relevant_submetrics(self, name: str, data: PreprocessedColumn) -> None:
+        if name not in self.submetrics:
+            return
+
+        for metric in self.submetrics[name].values():
+            metric.columnar_update(data)
+
+    def columnar_update(self, view: PreprocessedColumn) -> OperationResult:
+        count = 0
+        for value in list(chain.from_iterable(view.raw_iterator())):
```

Comment:
special case for the different vies and map/apply where possible

Reply:
Good idea for follow up, we can start with this as is.